File: Advanced/snakesandladders-cordapp/oracle-flows/src/main/java/net/corda/samples/snl/oracle/flows/DiceRollerFlow.java
package net.corda.samples.snl.oracle.flows;

import co.paralleluniverse.fibers.Suspendable;
import net.corda.core.flows.FlowException;
import net.corda.core.flows.FlowLogic;
import net.corda.core.flows.InitiatingFlow;
import net.corda.core.identity.CordaX500Name;
import net.corda.core.identity.Party;

@InitiatingFlow
public class DiceRollerFlow extends FlowLogic<Integer> {

    private String player;
    private Party oracle;

    public DiceRollerFlow(String player, Party oracle) {
        this.player = player;
        this.oracle = oracle;
    }

    @Override
    @Suspendable
    public Integer call() throws FlowException {
        return initiateFlow(oracle).sendAndReceive(Integer.class, player).unwrap(it->it);
    }
}

File: kotlin-samples/Advanced/snakesandladders-cordapp/oracle-flows/src/main/kotlin/net/corda/samples/snl/oracle/flows/DiceRollerFlow.kt
package net.corda.samples.snl.oracle.flows

import co.paralleluniverse.fibers.Suspendable
import net.corda.core.flows.*
import net.corda.core.identity.Party
import net.corda.core.utilities.unwrap
import net.corda.samples.snl.diceservice.DiceRollService


@InitiatingFlow
class DiceRollerFlow(private val player: String, private val oracle: Party) : FlowLogic<Int?>() {
    @Suspendable override fun call() = initiateFlow(oracle).sendAndReceive(Int::class.java, player).unwrap { it }
}


@InitiatedBy(DiceRollerFlow::class)
class DiceRollerFlowHandler(private val requestSession: FlowSession) : FlowLogic<Unit>() {
    @Suspendable
    @Throws(FlowException::class)
    override fun call(): Unit {
        val player = requestSession.receive(String::class.java).unwrap{ it -> it }
        val service: DiceRollService = serviceHub.cordaService(DiceRollService::class.java)
        val roll: Int = service.diceRoll(player)
        requestSession.send(roll)
        return
    }
}
